### PR TITLE
Error out when customizing a verity enabled base image.

### PR DESCRIPTION
**Pull request by liulanze**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

What does the PR accomplish, why was it needed?
It is possible customers may want to patch an existing image with verity enabled. The implementation should have some consistency and clearly calls out expectations - Options:
- Error out when that happens 
- Override the existing calculated verity hash and kernel args
We choose to achieve the first option erroring out when seeing the situation.

###### Does this affect the toolchain?  

**NO**

###### Associated issues  

- internal bug tracking: https://dev.azure.com/mariner-org/ECF/_workitems/edit/6830/

###### Test Methodology

- Pipeline build id: local MIC build.

### CODE REVIEW

File: toolkit/tools/pkg/imagecustomizerlib/partitionutils.go [hunk 1]
```diff
@@ -126,6 +126,14 @@ func findRootfsPartition(diskPartitions []diskutils.PartitionInfo, buildDir stri
 			continue
 		}
 
+		// Check if the partition is using DM_verity_hash file system type.
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/partitionutils.go:129]: Please move this check outside of the `findRootfsPartition` function.

Ideally, a function/class/etc should do 1 thing and do it well. You should avoid attaching extra behaviors to things unless there isn't any other good options. In this case, it is pretty trivial to have another `for` loop over the `diskPartitions` list.

File: toolkit/tools/pkg/imagecustomizerlib/partitionutils.go [hunk 2]
```diff
@@ -126,6 +126,14 @@ func findRootfsPartition(diskPartitions []diskutils.PartitionInfo, buildDir stri
...
+		// The presence of this type indicates that dm-verity has been enabled on the base image. If dm-verity is not enabled,
+		// the verity hash device should not be assigned this type. We do not support customization on verity enabled base
+		// images at this time because such modifications would compromise the integrity and security mechanisms enforced by dm-verity.
+		if diskPartition.FileSystemType == "DM_verity_hash" {
+			return nil, fmt.Errorf("this image has dm-verity pre-enabled, stop the customized build process")
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/partitionutils.go:134]: nit: "cannot customize base image that has dm-verity enabled"

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/pkg/imagecustomizerlib/partitionutils.go (hunk 1 above)
Comment (liulanze) [reply toolkit/tools/pkg/imagecustomizerlib/partitionutils.go]: Update the logic, please check if this logic makes sense.

**Review by romoh — APPROVED** [review]

File: toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go [hunk 3]
```diff
@@ -524,3 +533,27 @@ func customizeVerityImageHelper(buildDir string, baseConfigPath string, config *
...
 	return nil
 }
+
+func isDmVerityEnabled(buildDir string, rawImageFile string) error {
+	imageConnection := NewImageConnection()
+	err := imageConnection.ConnectLoopback(rawImageFile)
+	if err != nil {
+		return err
+	}
+	defer imageConnection.Close()
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go:543]: You should call `imageConnection.CleanClose` at the end of the function to ensure that the image file is no longer being used before continuing. (`imageConnection.Close` closes everything asynchronously.)

File: toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go [hunk 4]
```diff
@@ -524,3 +533,27 @@ func customizeVerityImageHelper(buildDir string, baseConfigPath string, config *
...
+	diskPartitions, err := diskutils.GetDiskPartitions(imageConnection.Loopback().DevicePath())
+	if err != nil {
+		return err
+	}
+
+	for i := range diskPartitions {
+		diskPartition := diskPartitions[i]
+
+		if diskPartition.FileSystemType == "DM_verity_hash" {
+			return fmt.Errorf("cannot customize base image that has dm-verity enabled.")
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go:554]: nit: Remove period (`.`)

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go (hunk 3 above)
Comment (liulanze) [reply toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go:543]: Fixed, plz check.

**Review by cwize1 — APPROVED** [review]

**Pull request merged**